src/pip/_internal/index/collector.py

```diff
@@ -558,6 +558,7 @@ def create(
         search_scope = SearchScope.create(
             find_links=find_links,
             index_urls=index_urls,
+            no_index=options.no_index,
```

Comment:
Does setting `index_urls` to an empty list when `no_index` is True achieve the same thing?

Reply:
The key bit is not here, but in `req_file.py`; there you need some mechanism to preserve the information that `no_index` was set at some point (either as part of the command line options or in any of the requirements files that will be recursively parsed). Over there you cannot simply set `index_urls=[]` because when you visit that point again you must know if the list was simply empty (and new entries can be added) or if it was cleared explicitly with `no_entry` and hence new entries may never be added. The mechanism that `req_file.py` uses to pass information around is the finder, in particular the `SearchScope`. Which is why we end up here.